pkg/segment/structs/metricsstructs.go

```diff
@@ -369,3 +388,20 @@ func (mbs *MBlockSummary) UpdateTimeRange(ts uint32) {
 		atomic.StoreUint32(&mbs.LowTs, ts)
 	}
 }
+
+func (metricFunc Function) Clone() *Function {
+	return &Function{
+		MathFunction:  metricFunc.MathFunction,
+		RangeFunction: metricFunc.RangeFunction,
+		ValueList:     metricFunc.ValueList,
+		TimeWindow:    metricFunc.TimeWindow,
+		Step:          metricFunc.Step,
+	}
+}
+
+func (agg Aggregation) Clone() *Aggregation {
+	return &Aggregation{
+		AggregatorFunction: agg.AggregatorFunction,
+		FuncConstant:       agg.FuncConstant,
+	}
+}
```

Comment:
Yes, I am using the clone function in this case here:

```
mQueryAgg := &structs.MetricQueryAgg{
		AggBlockType:    structs.AggregatorBlock,
		AggregatorBlock: mQuery.Aggregator.Clone(),
	}
```

I wanted to do clone because this AggregatorBlock is a pointer type but the Aggregator is not a pointer type but the direct value. 
So when I do `&Aggregation{}` and assigning these fields won't it automatically do something similar to deep clones, as all these are value types?

And I wanted it here because I will be giving a reference, in the next iteration if there is a change and that change is being applied here!! 

But are you saying instead of assigning them one by one... can I do something like this?
`return &Function{...metricFunc}`

Reply:
I'm not quite following. For `Aggregation` all the fields have primitive types, so a shallow copy and deep copy should be the same. For `Function`, I think the only field with a non-primitive type is `ValueList`, but you're doing a shallow copy of it anyways in your `Clone()` function. So I don't think there's any difference between:
```
copy1 := myFunction.Clone()
copy2 := myFunction
```
I'm pretty sure they'd both do a shallow copy. Maybe you meant to deep copy it (i.e., copy `ValuesList`)?

I don't think having `myFunction` as a pointer will change anything. See https://go.dev/play/p/B8xeDAXws72

Reply:
Actually I was wrong about the pointer bit, as that playground shows: https://go.dev/play/p/B8xeDAXws72. If that's the purpose of these `Clone()` methods, I think that's ok

Reply:
I think I get what you are trying to say, As you said I was not doing a deep copy. But that is not an issue within our code. As I am assigning the main `Aggregation` or `Function` to a new struct initialization. 

But as you said, we don't want to keep adding the new fields in the clone functions every time we update any of these structs. 

So I will keep the clone functions. But I will change them to this: 

```
func (metricFunc Function) Clone() *Function {
	functionCopy := metricFunc
	return &functionCopy
}

func (agg Aggregation) Clone() *Aggregation {
	aggCopy := agg
	return &aggCopy
}
```


Check this playground code once: https://go.dev/play/p/dyvSaF6fRuJ

Reply:
This code works, I have tested by running the test cases.

Reply:
That's a great idea. Also renaming these to `ShallowClone()` could add clarity. Or if `Clone()` is typically understood to mean a shallow copy, keeping it as `Clone()` works

Reply:
Yeah, I think ShallowClone makes it better. So in future it helps.